File: lib/sandbox/source_manifest.ml
type kind = Regular | Symlink

type entry = {
  path : string;
  kind : kind;
  executable : bool;
  size : int64;
  digest : string;
  target : string option;
  identity : string option;
}

type source =
  | Regular_source of {
      contents : string;
      executable : bool;
      identity : string option;
    }
  | Symlink_source of { target : string; identity : string option }

type exclusion = { path : string; reason : string }

type budget = {
  max_file_bytes : int;
  max_entries : int;
  max_snapshot_bytes : int64;
}

type t = {
  schema : string;
  entries : entry list;
  exclusions : exclusion list;
  exclusion_policy_digest : string;
  total_size : int64;
  canonical_json : string;
  digest : string;
}

let generated_directories =
  [
    ".git";
    ".workflow-verifier";
    ".workflow-verifier-cache";
    ".workflow-verifier-output";
  ]

let default_budget =
  {
    max_file_bytes = 16 * 1024 * 1024;
    max_entries = 100_000;
    max_snapshot_bytes = 4_294_967_296L;
  }

let normalize = Util.normalize_slashes

let normalized_root root =
  let root = normalize root in
  if root <> "." && root <> "" && Util.ends_with ~suffix:"/" root then
    String.sub root 0 (String.length root - 1)
  else root

let absolute_path path =
  Filename.is_relative path |> not
  || Util.starts_with ~prefix:"/" path
  || (String.length path >= 2 && path.[1] = ':')

let safe_segments path =
  path <> ""
  && (not (absolute_path path))
  && path |> String.split_on_char '/'
     |> List.for_all (fun segment ->
         segment <> "" && segment <> "." && segment <> "..")

let relative_to ~root path =
  let root = normalized_root root and path = normalize path in
  let rec strip_current_directory value =
    if Util.starts_with ~prefix:"./" value then
      strip_current_directory (String.sub value 2 (String.length value - 2))
    else value
  in
  let relative =
    if root = "." || root = "" then strip_current_directory path
    else if path = root then ""
    else
      let prefix = root ^ "/" in
      if Util.starts_with ~prefix path then
        String.sub path (String.length prefix)
          (String.length path - String.length prefix)
      else ""
  in
  if relative = "" then
    Error
      (Printf.sprintf "source path escapes or equals manifest root: %s" path)
  else if not (Util.valid_utf8 relative) then
    Error (Printf.sprintf "source path is not valid UTF-8: %s" path)
  else if not (safe_segments relative) then
    Error (Printf.sprintf "source path is not a safe relative path: %s" path)
  else Ok relative

let path_segments path = String.split_on_char '/' (String.lowercase_ascii path)

let excluded_by prefixes path =
  let lower = String.lowercase_ascii path in
  List.find_opt
    (fun prefix ->
      let prefix = String.lowercase_ascii (normalize prefix) in
      lower = prefix || Util.starts_with ~prefix:(prefix ^ "/") lower)
    prefixes

let generated_relative path =
  path_segments path
  |> List.exists (fun segment -> List.mem segment generated_directories)

let is_generated ~root path =
  match relative_to ~root path with
  | Ok relative -> generated_relative relative
  | Error _ -> false

let is_excluded ~root ~trusted_exclusions path =
  match relative_to ~root path with
  | Ok relative ->
      generated_relative relative
      || Option.is_some (excluded_by trusted_exclusions relative)
  | Error _ -> false

let kind_name = function
  | Regular -> "regular"
  | Symlink -> "symlink"

let entry_json (entry : entry) =
  Json.Object
    [
      ("digest", Json.String entry.digest);
      ("executable", Json.Bool entry.executable);
      ("kind", Json.String (kind_name entry.kind));
      ("path", Json.String entry.path);
      ("size", Json.Int64 entry.size);
      ( "target",
        Option.fold ~none:Json.Null
          ~some:(fun value -> Json.String value)
          entry.target );
    ]

let exclusion_json (exclusion : exclusion) =
  Json.Object
    [
      ("path", Json.String exclusion.path);
      ("reason", Json.String exclusion.reason);
    ]

let body_fields manifest =
  [
    ("entries", Json.Array (List.map entry_json manifest.entries));
    ("exclusion_policy_digest", Json.String manifest.exclusion_policy_digest);
    ("exclusions", Json.Array (List.map exclusion_json manifest.exclusions));
    ( "limits",
      Json.Object
        [
          ("max_entries", Json.Int default_budget.max_entries);
          ("max_file_bytes", Json.Int default_budget.max_file_bytes);
          ("max_snapshot_bytes", Json.Int64 default_budget.max_snapshot_bytes);
        ] );
    ("schema", Json.String manifest.schema);
    ("total_size", Json.Int64 manifest.total_size);
  ]

let body_json manifest = Json.Object (body_fields manifest)

let to_json manifest =
  Json.Object (("digest", Json.String manifest.digest) :: body_fields manifest)

let to_canonical_json manifest = Json.to_string (to_json manifest) ^ "\n"

let resolve_target path target =
  let target = normalize target in
  if absolute_path target then Error "absolute symlink target is forbidden"
  else if not (Util.valid_utf8 target) then
    Error "symlink target is not valid UTF-8"
  else
    let base = Filename.dirname path |> normalize in
    let rec fold stack = function
      | [] -> Ok (String.concat "/" (List.rev stack))
      | "" :: rest | "." :: rest -> fold stack rest
      | ".." :: rest -> (
          match stack with
          | [] -> Error "symlink target escapes snapshot root"
          | _ :: stack -> fold stack rest)
      | segment :: rest -> fold (segment :: stack) rest
    in
    fold []
      (String.split_on_char '/'
         (if base = "." then target else base ^ "/" ^ target))

let validate_symlinks entries =
  let targets =
    entries
    |> List.filter_map (fun entry ->
        match (entry.kind, entry.target) with
        | Symlink, Some target -> Some (entry.path, target)
        | _ -> None)
  in
  let rec visit visiting visited path =
    if List.mem path visiting then Error ("symlink cycle at " ^ path)
    else if List.mem path visited then Ok visited
    else
      match List.assoc_opt path targets with
      | None -> Ok (path :: visited)
      | Some target -> visit (path :: visiting) (path :: visited) target
  in
  let rec loop visited = function
    | [] -> Ok ()
    | (path, _) :: rest -> (
        match visit [] visited path with
        | Error _ as error -> error
        | Ok visited -> loop visited rest)
  in
  loop [] targets

let create_from_sources ~budget ~trusted_exclusions ~root ~files =
  if
    budget.max_file_bytes < default_budget.max_file_bytes
    || budget.max_entries < default_budget.max_entries
    || budget.max_snapshot_bytes < default_budget.max_snapshot_bytes
  then Error "source-manifest-v2 budgets are below the published floor"
  else
    let exclusion_policy =
      Json.Object
        [
          ( "default",
            Json.Array
              (List.map (fun value -> Json.String value) generated_directories)
          );
          ( "trusted",
            Json.Array
              (trusted_exclusions |> Util.deduplicate_strings
              |> List.map (fun value -> Json.String (normalize value))) );
        ]
      |> Json.to_string
    in
    let exclusion_policy_digest =
      "sha256:" ^ Sha256.digest_string exclusion_policy
    in
    let rec collect paths portable identities total entries exclusions =
      function
      | [] -> (
          let entries = List.rev entries and exclusions = List.rev exclusions in
          let manifest =
            {
              schema = "source-manifest-v2";
              entries;
              exclusions;
              exclusion_policy_digest;
              total_size = total;
              canonical_json = "";
              digest = "";
            }
          in
          let unsigned_json = Json.to_string (body_json manifest) in
          let digest = "sha256:" ^ Sha256.digest_string unsigned_json in
          let manifest = { manifest with canonical_json = ""; digest } in
          let manifest =
            { manifest with canonical_json = Json.to_string (to_json manifest) }
          in
          match validate_symlinks entries with
          | Error _ as error -> error
          | Ok () -> Ok manifest)
      | (path, source) :: rest -> (
          match relative_to ~root path with
          | Error _ as error -> error
          | Ok relative -> (
              match
                if generated_relative relative then Some "product-default"
                else
                  Option.map
                    (fun _ -> "trusted-policy")
                    (excluded_by trusted_exclusions relative)
              with
              | Some reason ->
                  collect paths portable identities total entries
                    ({ path = relative; reason } :: exclusions)
                    rest
              | None ->
                  let folded = String.lowercase_ascii relative in
                  if List.mem relative paths then
                    Error ("duplicate source manifest path: " ^ relative)
                  else if List.mem folded portable then
                    Error ("portable case-fold path collision: " ^ relative)
                  else if List.length entries >= budget.max_entries then
                    Error
                      "Incomplete.Resource_limit: source entry budget exceeded"
                  else
                    let kind, executable, size, digest, target, identity =
                      match source with
                      | Regular_source { contents; executable; identity } ->
                          ( Regular,
                            executable,
                            Int64.of_int (String.length contents),
                            "sha256:" ^ Sha256.digest_string contents,
                            None,
                            identity )
                      | Symlink_source { target; identity } -> (
                          match resolve_target relative target with
                          | Error message ->
                              raise
                                (Invalid_argument (relative ^ ": " ^ message))
                          | Ok resolved ->
                              ( Symlink,
                                false,
                                Int64.of_int (String.length target),
                                "sha256:" ^ Sha256.digest_string target,
                                Some resolved,
                                identity ))
                    in
                    if size > Int64.of_int budget.max_file_bytes then
                      Error
                        ("Incomplete.Resource_limit: file exceeds 16 MiB: "
                       ^ relative)
                    else if Int64.add total size > budget.max_snapshot_bytes
                    then
                      Error "Incomplete.Resource_limit: snapshot exceeds 4 GiB"
                    else if
                      Option.fold ~none:false
                        ~some:(fun value -> List.mem value identities)
                        identity
                    then Error ("hardlink/file identity collision: " ^ relative)
                    else
                      collect (relative :: paths) (folded :: portable)
                        (Option.to_list identity @ identities)
                        (Int64.add total size)
                        ({
                           path = relative;
                           kind;
                           executable;
                           size;
                           digest;
                           target;
                           identity;
                         }
                        :: entries)
                        exclusions rest))
    in
    try
      files
      |> List.sort (fun (left, _) (right, _) ->
          String.compare (normalize left) (normalize right))
      |> collect [] [] [] 0L [] []
    with Invalid_argument message -> Error message

let create ~root ~files =
  let files =
    files
    |> List.map (fun (path, contents) ->
        (path, Regular_source { contents; executable = false; identity = None }))
  in
  create_from_sources ~budget:default_budget ~trusted_exclusions:[] ~root ~files

File: crates/sandbox/src/source_manifest.rs
use std::collections::{BTreeMap, BTreeSet};
use workflow_verifier_foundation::{
    JsonValue, PublicPath, content_digest, normalize_slashes, portable_path_key,
};

const GENERATED_DIRECTORIES: [&str; 4] = [
    ".git",
    ".workflow-verifier",
    ".workflow-verifier-cache",
    ".workflow-verifier-output",
];

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct ManifestBudget {
    pub max_file_bytes: u64,
    pub max_entries: usize,
    pub max_snapshot_bytes: u64,
}

impl Default for ManifestBudget {
    fn default() -> Self {
        Self {
            max_file_bytes: 16 * 1024 * 1024,
            max_entries: 100_000,
            max_snapshot_bytes: 4_294_967_296,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
enum SourceKind {
    Regular { contents: Vec<u8>, executable: bool },
    Symlink { target: String },
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SourceFile {
    path: String,
    kind: SourceKind,
    identity: Option<String>,
}

impl SourceFile {
    #[must_use]
    pub fn regular(path: impl Into<String>, contents: impl AsRef<[u8]>) -> Self {
        Self {
            path: path.into(),
            kind: SourceKind::Regular {
                contents: contents.as_ref().to_vec(),
                executable: false,
            },
            identity: None,
        }
    }

    #[must_use]
    pub fn executable(path: impl Into<String>, contents: impl AsRef<[u8]>) -> Self {
        Self {
            path: path.into(),
            kind: SourceKind::Regular {
                contents: contents.as_ref().to_vec(),
                executable: true,
            },
            identity: None,
        }
    }

    #[must_use]
    pub fn symlink(path: impl Into<String>, target: impl Into<String>) -> Self {
        Self {
            path: path.into(),
            kind: SourceKind::Symlink {
                target: target.into(),
            },
            identity: None,
        }
    }

    #[must_use]
    pub fn with_identity(mut self, identity: impl Into<String>) -> Self {
        self.identity = Some(identity.into());
        self
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ManifestEntryKind {
    Regular,
    Symlink,
}

impl ManifestEntryKind {
    fn name(self) -> &'static str {
        match self {
            Self::Regular => "regular",
            Self::Symlink => "symlink",
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ManifestEntry {
    pub path: String,
    pub kind: ManifestEntryKind,
    pub executable: bool,
    pub size: u64,
    pub digest: String,
    pub target: Option<String>,
    pub identity: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ManifestExclusion {
    pub path: String,
    pub reason: String,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct SourceManifest {
    pub entries: Vec<ManifestEntry>,
    pub exclusions: Vec<ManifestExclusion>,
    pub exclusion_policy_digest: String,
    pub total_size: u64,
    pub digest: String,
}

impl SourceManifest {
    /// Create the immutable source-manifest-v2 protocol object.
    ///
    /// # Errors
    /// Rejects unsafe paths and links, collisions, hardlink aliases, and resource excess.
    pub fn create(
        root: &str,
        files: impl IntoIterator<Item = SourceFile>,
        trusted_exclusions: &[String],
    ) -> Result<Self, String> {
        Self::create_with_budget(root, files, trusted_exclusions, ManifestBudget::default())
    }

    /// Create a manifest under an explicit, non-widening resource envelope.
    ///
    /// # Errors
    /// Returns a stable incomplete or validation reason for the first violation.
    #[allow(clippy::too_many_lines)]
    pub fn create_with_budget(
        root: &str,
        files: impl IntoIterator<Item = SourceFile>,
        trusted_exclusions: &[String],
        budget: ManifestBudget,
    ) -> Result<Self, String> {
        let published = ManifestBudget::default();
        if budget.max_file_bytes < published.max_file_bytes
            || budget.max_entries < published.max_entries
            || budget.max_snapshot_bytes < published.max_snapshot_bytes
        {
            return Err("source-manifest-v2 budgets are below the published floor".to_owned());
        }
        let root = normalize_root(root);
        let mut files: Vec<_> = files.into_iter().collect();
        files.sort_by(|left, right| {
            normalize_slashes(&left.path).cmp(&normalize_slashes(&right.path))
        });
        let trusted: BTreeSet<String> = trusted_exclusions
            .iter()
            .map(|path| normalize_slashes(path))
            .collect();
        let exclusion_policy = JsonValue::Object(BTreeMap::from([
            (
                "default".to_owned(),
                JsonValue::Array(
                    GENERATED_DIRECTORIES
                        .iter()
                        .map(|path| JsonValue::String((*path).to_owned()))
                        .collect(),
                ),
            ),
            (
                "trusted".to_owned(),
                JsonValue::Array(trusted.iter().cloned().map(JsonValue::String).collect()),
            ),
        ]));
        let exclusion_policy_digest = content_digest(exclusion_policy.canonical());
        let mut entries = Vec::new();
        let mut exclusions = Vec::new();
        let mut exact_paths = BTreeSet::new();
        let mut portable_paths = BTreeSet::new();
        let mut identities = BTreeSet::new();
        let mut total_size = 0u64;
        for file in files {
            let relative = relative_to(&root, &file.path)?;
            if generated(&relative) {
                exclusions.push(ManifestExclusion {
                    path: relative,
                    reason: "product-default".to_owned(),
                });
                continue;
            }
            if trusted.iter().any(|prefix| path_below(&relative, prefix)) {
                exclusions.push(ManifestExclusion {
                    path: relative,
                    reason: "trusted-policy".to_owned(),
                });
                continue;
            }
            if !exact_paths.insert(relative.clone()) {
                return Err(format!("duplicate source manifest path: {relative}"));
            }
            if !portable_paths.insert(portable_path_key(&relative)) {
                return Err(format!("portable case-fold path collision: {relative}"));
            }
            if entries.len() >= budget.max_entries {
                return Err("Incomplete.Resource_limit: source entry budget exceeded".to_owned());
            }
            if let Some(identity) = &file.identity
                && !identities.insert(identity.clone())
            {
                return Err(format!("hardlink/file identity collision: {relative}"));
            }
            let (kind, executable, bytes, target) = match file.kind {
                SourceKind::Regular {
                    contents,
                    executable,
                } => (ManifestEntryKind::Regular, executable, contents, None),
                SourceKind::Symlink { target } => {
                    let normalized = resolve_target(&relative, &target)?;
                    (
                        ManifestEntryKind::Symlink,
                        false,
                        target.as_bytes().to_vec(),
                        Some(normalized),
                    )
                }
            };
            let size = u64::try_from(bytes.len()).map_err(|_| "source file is too large")?;
            if size > budget.max_file_bytes {
                return Err(format!(
                    "Incomplete.Resource_limit: file exceeds 16 MiB: {relative}"
                ));
            }
            total_size = total_size
                .checked_add(size)
                .ok_or_else(|| "Incomplete.Resource_limit: snapshot size overflow".to_owned())?;
            if total_size > budget.max_snapshot_bytes {
                return Err("Incomplete.Resource_limit: snapshot exceeds 4 GiB".to_owned());
            }
            entries.push(ManifestEntry {
                path: relative,
                kind,
                executable,
                size,
                digest: content_digest(bytes),
                target,
                identity: file.identity,
            });
        }
        validate_symlinks(&entries)?;
        let mut manifest = Self {
            entries,
            exclusions,
            exclusion_policy_digest,
            total_size,
            digest: String::new(),
        };
        manifest.digest = content_digest(manifest.body_json().canonical());
        Ok(manifest)
    }

    fn body_fields(&self) -> BTreeMap<String, JsonValue> {
        BTreeMap::from([
            (
                "entries".to_owned(),
                JsonValue::Array(self.entries.iter().map(entry_json).collect()),
            ),
            (
                "exclusion_policy_digest".to_owned(),
                JsonValue::String(self.exclusion_policy_digest.clone()),
            ),
            (
                "exclusions".to_owned(),
                JsonValue::Array(self.exclusions.iter().map(exclusion_json).collect()),
            ),
            ("limits".to_owned(), limits_json()),
            (
                "schema".to_owned(),
                JsonValue::String("source-manifest-v2".to_owned()),
            ),
            (
                "total_size".to_owned(),
                JsonValue::Integer(i64::try_from(self.total_size).unwrap_or(i64::MAX)),
            ),
        ])
    }

    fn body_json(&self) -> JsonValue {
        JsonValue::Object(self.body_fields())
    }

    #[must_use]
    pub fn verify_digest(&self) -> bool {
        self.digest == content_digest(self.body_json().canonical())
    }

    #[must_use]
    pub fn to_canonical_json(&self) -> String {
        let mut fields = self.body_fields();
        fields.insert("digest".to_owned(), JsonValue::String(self.digest.clone()));
        JsonValue::Object(fields).canonical_line()
    }
}

fn entry_json(entry: &ManifestEntry) -> JsonValue {
    JsonValue::Object(BTreeMap::from([
        ("digest".to_owned(), JsonValue::String(entry.digest.clone())),
        (
            "executable".to_owned(),
            JsonValue::Boolean(entry.executable),
        ),
        (
            "kind".to_owned(),
            JsonValue::String(entry.kind.name().to_owned()),
        ),
        ("path".to_owned(), JsonValue::String(entry.path.clone())),
        (
            "size".to_owned(),
            JsonValue::Integer(i64::try_from(entry.size).unwrap_or(i64::MAX)),
        ),
        (
            "target".to_owned(),
            entry
                .target
                .clone()
                .map_or(JsonValue::Null, JsonValue::String),
        ),
    ]))
}

fn exclusion_json(exclusion: &ManifestExclusion) -> JsonValue {
    JsonValue::Object(BTreeMap::from([
        ("path".to_owned(), JsonValue::String(exclusion.path.clone())),
        (
            "reason".to_owned(),
            JsonValue::String(exclusion.reason.clone()),
        ),
    ]))
}

fn limits_json() -> JsonValue {
    let limits = ManifestBudget::default();
    JsonValue::Object(BTreeMap::from([
        (
            "max_entries".to_owned(),
            JsonValue::Integer(i64::try_from(limits.max_entries).unwrap_or(i64::MAX)),
        ),
        (
            "max_file_bytes".to_owned(),
            JsonValue::Integer(i64::try_from(limits.max_file_bytes).unwrap_or(i64::MAX)),
        ),
        (
            "max_snapshot_bytes".to_owned(),
            JsonValue::Integer(i64::try_from(limits.max_snapshot_bytes).unwrap_or(i64::MAX)),
        ),
    ]))
}

fn normalize_root(root: &str) -> String {
    let root = normalize_slashes(root);
    root.trim_end_matches('/').to_owned()
}

fn relative_to(root: &str, path: &str) -> Result<String, String> {
    let path = normalize_slashes(path);
    let path = path.trim_start_matches("./").to_owned();
    let relative = if matches!(root, "" | ".") {
        path
    } else if let Some(relative) = path.strip_prefix(&format!("{root}/")) {
        relative.to_owned()
    } else {
        return Err(format!(
            "source path escapes or equals manifest root: {path}"
        ));
    };
    PublicPath::new(relative.clone())
        .map_err(|_| format!("source path is not a safe relative path: {relative}"))?;
    Ok(relative)
}

fn generated(path: &str) -> bool {
    path.split('/').any(|segment| {
        GENERATED_DIRECTORIES
            .iter()
            .any(|generated| segment.eq_ignore_ascii_case(generated))
    })
}

fn path_below(path: &str, prefix: &str) -> bool {
    path.eq_ignore_ascii_case(prefix)
        || path
            .to_ascii_lowercase()
            .starts_with(&format!("{}/", prefix.to_ascii_lowercase()))
}

fn resolve_target(path: &str, target: &str) -> Result<String, String> {
    let target = normalize_slashes(target);
    if target.starts_with('/') || target.as_bytes().get(1) == Some(&b':') || target.contains('\0') {
        return Err(format!("{path}: absolute symlink target is forbidden"));
    }
    let mut segments: Vec<&str> = path
        .rsplit_once('/')
        .map_or(Vec::new(), |(base, _)| base.split('/').collect());
    for segment in target.split('/') {
        match segment {
            "" | "." => {}
            ".." => {
                if segments.pop().is_none() {
                    return Err(format!("{path}: symlink target escapes snapshot root"));
                }
            }
            other => segments.push(other),
        }
    }
    let resolved = segments.join("/");
    PublicPath::new(resolved.clone())
        .map_err(|_| format!("{path}: symlink target is not a safe relative path"))?;
    Ok(resolved)
}

fn validate_symlinks(entries: &[ManifestEntry]) -> Result<(), String> {
    let targets: BTreeMap<&str, &str> = entries
        .iter()
        .filter_map(|entry| {
            entry
                .target
                .as_deref()
                .map(|target| (entry.path.as_str(), target))
        })
        .collect();
    for start in targets.keys() {
        let mut visiting = BTreeSet::new();
        let mut cursor = *start;
        while let Some(target) = targets.get(cursor) {
            if !visiting.insert(cursor) {
                return Err(format!("symlink cycle at {cursor}"));
            }
            cursor = target;
        }
    }
    Ok(())
}
